lib/msf/core/post/windows/runas.rb

```diff
@@ -34,4 +40,204 @@ def shell_exec(command, args)
+                                                                      nil,
+                                                                      nil,
+                                                                      startup_info,
+                                                                      16)
+
+        if create_process['return']
+          begin
+            pi = parse_process_information(create_process['lpProcessInformation'])
+          ensure
+            session.railgun.kernel32.CloseHandle(pi[:process_handle])
+            session.railgun.kernel32.CloseHandle(pi[:thread_handle])
+          end
+          print_good("Process started successfully, PID: #{pi[:process_id]}")
+        else
+          print_error("Unable to create process, Error Code: #{create_process['GetLastError']} - #{create_process['ErrorMessage']}")
+        end
+
+        return pi
+      ensure
+        session.railgun.kernel32.CloseHandle(ph_token)
+      end
+    else
+      print_error("Unable to login the user, Error Code: #{logon_user['GetLastError']} - #{logon_user['ErrorMessage']}")
+    end
+
+    nil
+  end
+
+  #
+  # Parse the PROCESS_INFORMATION struct
+  #
+  # @param process_information [String] The PROCESS_INFORMATION value
+  #   from the CreateProcess call
+  #
+  # @return [Hash] The values from the process_information struct
+  #
+  def parse_process_information(process_information)
+    pi = process_information.unpack('LLLL')
+    { :process_handle => pi[0], :thread_handle => pi[1], :process_id => pi[2], :thread_id => pi[3] }
+  end
+
+  #
+  # Checks the username and domain is in the correct format
+  # for the CreateProcess_x WinAPI calls.
+  #
+  # @param username [String] The target user
+  # @param domain [String] The target user domain
+  #
+  # @raise [ArgumentError] If the username format is incorrect
+  #
+  # @return [True] True if username is in the correct format
+  #
+  def check_user_format(username, domain)
+    if domain && username.include?('@')
+      raise ArgumentError, 'Username is in UPN format (user@domain) so the domain parameter must be nil'
+    end
+
+    true
+  end
+
+  #
+  # Checks the command_length parameter is the correct length
+  # for the CreateProcess_x WinAPI calls depending on the presence
+  # of application_name
+  #
+  # @param application_name [String] lpApplicationName
+  # @param command_line [String] lpCommandLine
+  # @param max_length [Integer] The max command length of the respective
+  #   CreateProcess function
+  #
+  # @raise [ArgumentError] If the command_line is too large
+  #
+  # @return [True] True if the command_line is within the correct bounds
+  #
+  def check_command_length(application_name, command_line, max_length)
+    if application_name.nil? && command_line.nil?
+      raise ArgumentError, 'Both application_name and command_line are nil'
+    elsif application_name.nil? && command_line && command_line.length > MAX_PATH
```

Comment:
Think this is invalid, the wording from MSDN, think MAX_PATH only occurs to the module path (ie argv[0])


`The maximum length of this string is 32K characters. If lpApplicationName is NULL, the module name portion of lpCommandLine is limited to MAX_PATH characters.`